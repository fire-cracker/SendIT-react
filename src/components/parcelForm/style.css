 /*Styling for the body background*/
 body {
  margin: 0px;
  width: 100%;
  height: 100%;
  padding: 0px;
}

/* Styling the create Shipment Section */
.shipmentOrder{
  background-image: linear-gradient(rgba(0, 0, 0, .5), rgba(0, 0, 0, .5)), url("../../../public/img/background3.jpg");
  background-position: center;
  background-size: cover;
  height: 100%;
  width:100%;
  min-height: 100vh;
  padding-top: 10%;
}

.shipmentOrder>h1{
  text-align: center;
  color:white;
}

.createShipment{
 width: 100%;
}

.fromform{
  margin:auto;
  width: 60%;
  overflow: auto;
  padding: 18px 4% 20px 4%;
  border: 1px solid #888;
  box-shadow:	
      0pt 2px 5px rgba(105, 108, 109,  0.7),	
      0px 0px 8px 5px rgba(208, 223, 226, 0.4) inset;
  border-radius: 5px;	
 
}

.createShipmentParticulars{
  padding: 10px;
}

.createShipmentParticulars>span{
 color:white;
}

.createShipmentParticulars>input{
color: white;
width: 100%;
padding: 5px 20px;
display: inline-block;
box-sizing: border-box;
outline:none;
font-size: 120%;
background: transparent;
}


/* Turn off number input spinners */
input[type=number]::-webkit-inner-spin-button, 
input[type=number]::-webkit-outer-spin-button { 
 -webkit-appearance: none; 
 margin: 0; 
}

/* Set a style for all buttons */

.createShipmentButton>button:hover {
  opacity: 0.8;
}
.createShipmentButton>button {
  padding: 1% 3%;
  border-radius: 10px;
  font-size: 120%;
  float: right;
  }
 /* For desktop: */
 @media only screen and (min-width: 768px) {
  /* .fromform{
     width: 40%;
  } */
  .fromform{
    margin:auto;
    width: 35%;
  }
}