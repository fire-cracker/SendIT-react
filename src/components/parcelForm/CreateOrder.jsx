import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { toast } from 'react-toastify';

import { addParcel } from '../../redux/actions/createOrder';
import Header from '../header';
import Button from '../Button';
import './style.css';

export class CreateOrder extends Component {
  state = {
    pickupLocation: '',
    destination: '',
    weight: '',
    pickupLocationErrorMessage: '',
    destinationErrorMessage: '',
    weightErrorMessage: '',
  }

  componentDidUpdate() {
    const { history, login: { isLoggedIn } } = this.props;
    if (isLoggedIn === false) {
      history.push('/');
    }
  }

OnChangePickupLocation = (e) => {
  this.setState({
    pickupLocation: e.target.value,
    pickupLocationErrorMessage: ''
  });
}

onChangeDestination = (e) => {
  this.setState({
    destination: e.target.value,
    destinationErrorMessage: ''
  });
}

onChangeWeight = (e) => {
  this.setState({
    weight: e.target.value,
    weightErrorMessage: ''
  });
}

onPickupLocationBlur = () => {
  const { pickupLocation } = this.state;

  if (!pickupLocation) {
    return this.setState({
      pickupLocationErrorMessage: 'Please enter the PickupLocation of your parcel.'
    });
  }

  if (pickupLocation.length < 8 || pickupLocation.length > 50) {
    return this.setState({
      pickupLocationErrorMessage: 'PickupLocation must between 8 and 50 characters.'
    });
  }
};

onDestinationBlur = () => {
  const { destination } = this.state;

  if (!destination) {
    return this.setState({
      destinationErrorMessage: 'Please enter the Destination of your parcel.'
    });
  }

  if (destination.length < 8 || destination.length > 50) {
    return this.setState({
      destinationErrorMessage: 'Destination must between 8 and 50 characters.'
    });
  }
};

onWeightBlur = () => {
  const { weight } = this.state;

  if (!weight) {
    return this.setState({
      weightErrorMessage: 'Please enter the weight of your parcel.'
    });
  }

  if (weight > 20) {
    return this.setState({
      weightErrorMessage: 'Parcel must not be more than 20kg.'
    });
  }
};

hasFormErrors = () => {
  const {
    pickupLocationErrorMessage,
    destinationErrorMessage,
    weightErrorMessage,
  } = this.state;

  if (pickupLocationErrorMessage || destinationErrorMessage || weightErrorMessage) {
    return true;
  }
  return false;
}

hasEmptyRequiredFields = () => {
  const {
    pickupLocation,
    destination,
    weight,
  } = this.state;

  if (!pickupLocation || !destination || !weight) {
    return true;
  }
  return false;
}

onClickSubmit = (e) => {
  e.preventDefault();
  const { history } = this.props;
  const { pickupLocation, destination, weight } = this.state;
  this.props.addParcel({ pickupLocation, destination, weight })
    .then()
    .catch((error) => {
      const { response, response: { status, data } } = error;
      if (response && status === 400) {
        return toast.error(data.message[0]);
      }
      if (response && status === 403) {
        return toast.error('You need to login to create an article');
      }
      if (response && status === 401) {
        history.push('/');
        return toast.error('Your session has expired. You need to login');
      }
      toast.error('Unknown error');
    });
  this.setState({
    pickupLocation: '',
    destination: '',
    weight: ''
  });
}

render() {
  const {
    pickupLocation,
    destination,
    weight,
    pickupLocationErrorMessage,
    destinationErrorMessage,
    weightErrorMessage
  } = this.state;

  const buttonDisabled = this.hasFormErrors() || this.hasEmptyRequiredFields();

  return (
    <div className="createOrder">
      <div>
        <Header />
      </div>
      <div className="shipmentOrder">
        <h1> Create Shipment </h1>
        <div className="createShipment">
          <form className="fromform" id="fromform" autoComplete="on">

            <div className="createShipmentParticulars">
              <span className="addressLabel"><b>Pick Location</b></span>
              <small className="password-error-message">
            &nbsp;
                {pickupLocationErrorMessage}
              </small>
              <input
                type="text"
                id="pickLocation"
                name="address"
                value={pickupLocation}
                onChange={this.OnChangePickupLocation}
                onBlur={this.onPickupLocationBlur}
              />
              <span className="emailLabel"><b>Destination</b></span>
              <small className="password-error-message">
            &nbsp;
                {destinationErrorMessage}
              </small>
              <input
                type="text"
                id="destination"
                name="destination"
                value={destination}
                onChange={this.onChangeDestination}
                onBlur={this.onDestinationBlur}
              />
              <span className="weightLabel"><b>weight(kg) </b></span>
              <small className="password-error-message">
            &nbsp;
                {weightErrorMessage}
              </small>
              <input
                type="number"
                id="weight"
                name="weight"
                value={weight}
                onChange={this.onChangeWeight}
                onBlur={this.onWeightBlur}
              />
            </div>
            <div className="createShipmentButton">
              <Button
                type="submit"
                className="shipmentButton"
                onClick={this.onClickSubmit}
                value="submit"
                disabled={buttonDisabled}
              />
            </div>
          </form>


        </div>


      </div>
    </div>
  );
}
}

CreateOrder.propTypes = {
  addParcel: PropTypes.func,
  history: PropTypes.object,
  login: PropTypes.object,
};
const mapStateToProps = ({ login }) => ({ login });

const mapDispatchToProps = ({
  addParcel
});

export default connect(mapStateToProps, mapDispatchToProps)(CreateOrder);
