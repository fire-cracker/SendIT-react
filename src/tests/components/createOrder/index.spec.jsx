import React from 'react';
import { shallow } from 'enzyme';
import { toast } from 'react-toastify';
import axios from '../../../utils/axiosConfig';

import { CreateOrder } from '../../../components/parcelForm/CreateOrder';

jest.mock('../../../utils/axiosConfig');
jest.mock('react-toastify');
jest.mock('../../../redux/actions/login');
jest.mock('../../../redux/actions/signup');


let wrapper;
const preventDefault = jest.fn();

describe('Tests for the Create Order Form Page', () => {
  beforeEach(() => {
    const props = {
      addParcel: jest.fn().mockImplementation(() => Promise.resolve({})),
      login: {
        isLoggedIn: true
      }
    };
    wrapper = shallow(<CreateOrder {...props} />);
  });

  it('Should render the Create Order Form page', () => {
    expect(wrapper.length).toBe(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('should set state onChange', () => {
    wrapper.find('input').at(0).simulate('change', {
      target: { value: 'm' }
    });
    expect(wrapper.state('pickupLocation')).toBe('m');

    wrapper.find('input').at(1).simulate('change', {
      target: { value: 'm' }
    });
    expect(wrapper.state('destination')).toBe('m');

    wrapper.find('input').at(2).simulate('change', {
      target: { value: 'm' }
    });
    expect(wrapper.state('weight')).toBe('m');
  });

  it('should test for onBlur', () => {
    wrapper.find('input').at(0).simulate('change', {
      target: { value: '' }
    });
    wrapper.find('input').at(0).simulate('blur');
    expect(wrapper.state('pickupLocationErrorMessage')).toBe('Please enter the PickupLocation of your parcel.');

    wrapper.find('input').at(0).simulate('change', {
      target: {
        value: "Just like the '90s, good things have to come to an end. If you want to switch to the current site, you'll find the option here."
      }
    });
    wrapper.find('input').at(0).simulate('blur');
    expect(wrapper.state('pickupLocationErrorMessage')).toBe(
      'PickupLocation must between 8 and 50 characters.'
    );

    wrapper.find('input').at(1).simulate('change', {
      target: { value: '' }
    });
    wrapper.find('input').at(1).simulate('blur');
    expect(wrapper.state('destinationErrorMessage')).toBe(
      'Please enter the Destination of your parcel.'
    );

    wrapper.find('input').at(1).simulate('change', {
      target: {
        value: "Just like the '90s, good things have to come to an end. If you want to switch to the current site, you'll find the option here.Just like the '90s, good things have to come to an end."
      }
    });
    wrapper.find('input').at(1).simulate('blur');
    expect(wrapper.state('destinationErrorMessage')).toBe(
      'Destination must between 8 and 50 characters.'
    );

    wrapper.find('input').at(2).simulate('change', {
      target: { value: '' }
    });
    wrapper.find('input').at(2).simulate('blur');
    expect(wrapper.state('weightErrorMessage')).toBe(
      'Please enter the weight of your parcel.'
    );

    wrapper.find('input').at(2).simulate('change', {
      target: { value: 30 }
    });
    wrapper.find('input').at(2).simulate('blur');
    expect(wrapper.state('weightErrorMessage')).toBe(
      'Parcel must not be more than 20kg.'
    );
  });

  it('should create order if correct order details is provided', async () => {
    window.history.pushState({}, '/');
    const props = {
      addParcel: jest.fn().mockImplementation(() => Promise.resolve({})),
      login: {
        isLoggedIn: true
      }
    };
    wrapper = shallow(<CreateOrder {...props} />);
    wrapper.find('input').at(0).simulate('change', {
      target: { value: 'Oyedeji Haven, Jagun Old Ife Road Ibadan' }
    });

    wrapper.find('input').at(1).simulate('change', {
      target: { value: 'Oyedeji Haven, Jagun Old Ife Road Ibadan' }
    });

    wrapper.find('input').at(2).simulate('change', {
      target: { value: 2 }
    });

    wrapper.find('Button').simulate('click', { preventDefault });
    expect(props.addParcel).toHaveBeenCalled();
  });


  it('should throw an error if correct order details are incorrect', async () => {
    const props = {
      addParcel: jest.fn().mockImplementation(() => Promise.reject({})),
      login: {
        isLoggedIn: true
      }
    };
    wrapper = shallow(<CreateOrder {...props} />);

    wrapper.find('Button').simulate('click', { preventDefault });
    axios.post.mockResolvedValue(props.addParcel);
  });

  it('should throw toast error if error code is 400', async () => {
    toast.error = jest.fn();
    const props = {
      addParcel: jest.fn().mockImplementation(() => Promise.reject({
        response: {
          status: 400
        }
      })),
      login: {
        isLoggedIn: true
      }
    };
    wrapper = shallow(<CreateOrder {...props} />);

    wrapper.find('Button').simulate('click', { preventDefault });
    axios.post.mockResolvedValue(props.addParcel);
  });

  it('should throw toast error if error code is 403', async () => {
    toast.error = jest.fn();
    const props = {
      addParcel: jest.fn().mockImplementation(() => Promise.reject({
        response: {
          status: 403
        }
      })),
      login: {
        isLoggedIn: true
      }
    };
    wrapper = shallow(<CreateOrder {...props} />);

    wrapper.find('Button').simulate('click', { preventDefault });
    axios.post.mockResolvedValue(props.addParcel);
  });
  it('should throw toast error if error code is 401', async () => {
    toast.error = jest.fn();
    const props = {
      addParcel: jest.fn().mockImplementation(() => Promise.reject({
        response: {
          status: 401
        }
      })),
      login: {
        isLoggedIn: true
      }
    };
    wrapper = shallow(<CreateOrder {...props} />);

    wrapper.find('Button').simulate('click', { preventDefault });
    axios.post.mockResolvedValue(props.addParcel);
  });
  it('should throw toast error if error code is 401', async () => {
    toast.error = jest.fn();
    const props = {
      addParcel: jest.fn().mockImplementation(() => Promise.reject({
        response: {
          status: 500
        }
      })),
      login: {
        isLoggedIn: true
      }
    };
    wrapper = shallow(<CreateOrder {...props} />);

    wrapper.find('Button').simulate('click', { preventDefault });
    axios.post.mockResolvedValue(props.addParcel);
  });
});
