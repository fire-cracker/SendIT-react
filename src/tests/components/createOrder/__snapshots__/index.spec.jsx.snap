// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for the Create Order Form Page Should render the Create Order Form page 1`] = `
<div
  className="createOrder"
>
  <div>
    <Header />
  </div>
  <div
    className="shipmentOrder"
  >
    <h1>
       Create Shipment 
    </h1>
    <div
      className="createShipment"
    >
      <form
        autoComplete="on"
        className="fromform"
        id="fromform"
      >
        <div
          className="createShipmentParticulars"
        >
          <span
            className="addressLabel"
          >
            <b>
              Pick Location
            </b>
          </span>
          <small
            className="password-error-message"
          >
             
          </small>
          <input
            id="pickLocation"
            name="address"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value=""
          />
          <span
            className="emailLabel"
          >
            <b>
              Destination
            </b>
          </span>
          <small
            className="password-error-message"
          >
             
          </small>
          <input
            id="destination"
            name="destination"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value=""
          />
          <span
            className="weightLabel"
          >
            <b>
              weight(kg) 
            </b>
          </span>
          <small
            className="password-error-message"
          >
             
          </small>
          <input
            id="weight"
            name="weight"
            onBlur={[Function]}
            onChange={[Function]}
            type="number"
            value=""
          />
        </div>
        <div
          className="createShipmentButton"
        >
          <Button
            className="shipmentButton"
            disabled={true}
            onClick={[Function]}
            type="submit"
            value="submit"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
